src/celeritas/track/detail/TrackSortUtils.cu

```diff
@@ -51,6 +53,33 @@ void shuffle_track_slots<MemSpace::device>(
         g);
     CELER_DEVICE_CHECK_ERROR();
 }
+
+namespace
+{
+struct alive_predicate
+{
+    CELER_FUNCTION bool operator()(TrackStatus const& track_status) const
+    {
+        return track_status == TrackStatus::alive;
+    }
+};
+}  // namespace
+
+template<>
+void partition_tracks_by_status(
+    CoreStateData<Ownership::reference, MemSpace::device> const& states)
+{
+    CELER_EXPECT(states.size() > 0);
+    Span track_slots{
+        states.track_slots[AllItems<TrackSlotId::size_type, MemSpace::device>{}]};
+    Span status{states.sim.status[AllItems<TrackStatus, MemSpace::device>{}]};
+    thrust::partition(thrust::device,
+                      thrust::device_pointer_cast(track_slots.begin()),
+                      thrust::device_pointer_cast(track_slots.end()),
+                      thrust::device_pointer_cast(status.data()),
+                      alive_predicate{});
```

Comment:
Actually, I think this implementation is not correct. `status` is indexed by `TrackSlotId` and `track_slots` by `ThreadId`, so the track slots partitioned together aren't going to redirect to alive tracks unless the condition `track_slots[i] == i` which makes all `ThreadId` and `TrackSlotId` equivalent.

Reply:
Two possible implementations: 

- "Reset" `track_slots by  calling `fill_track_slots` before partitioning.
- Capture status in the predicate as done in 705dee5

I don't think there is a large performance difference between the two options, I'd say the 2nd option is best because we have a single kernel launch.